DMCompiler/DM/Visitors/DMProcBuilder.cs

```diff
@@ -451,6 +460,55 @@ public void ProcessStatementSet(DMASTProcStatementSet statementSet)
             _proc.DestroyEnumerator();
         }
 
+        public void ProcessStatementForType(DMASTProcStatementForType statementForType)
+        {
+            DMASTProcStatementVarDeclaration varDeclaration = statementForType.Initializer as DMASTProcStatementVarDeclaration;
+
+            if (varDeclaration?.Type != null)
+            {
+                _proc.PushPath(varDeclaration.Type.Value);
+                _proc.CreateTypeEnumerator();
+            }
+            else // This shouldn't happen, just to be safe
+            {
+                DMCompiler.Error(new CompilerError(varDeclaration.Location, "Attempted to create a type enumerator with a null type"));
+                return;
+            }
+
+            _proc.StartScope();
+            {
+                if (statementForType.Initializer != null) {
+                    ProcessStatement(statementForType.Initializer);
+                }
+
+                string loopLabel = _proc.NewLabelName();
+                _proc.LoopStart(loopLabel);
+                {
+                    DMExpression outputVariable = DMExpression.Create(_dmObject, _proc, statementForType.Variable);
+                    (DMReference outputRef, _) = outputVariable.EmitReference(_dmObject, _proc);
+                    _proc.Enumerate(outputRef);
+                    _proc.BreakIfFalse();
+
+                    if (varDeclaration != null && varDeclaration.Type != null)
+                    {
+                        DMExpression.Emit(_dmObject, _proc, statementForType.Variable);
+                        _proc.PushPath(varDeclaration.Type.Value);
+                        _proc.IsType();
+
+                        _proc.ContinueIfFalse();
+                    }
```

Comment:
Having `CreateTypeEnumerator` only enumerate its given type instead of the more general `world.contents` or `state.DreamManager.Datums` would allow this `istype()` check to be removed